### ISSUE TITLE
No value space checking of numeric field values

### ISSUE THREAD

**Issue by martinjoconnor** [metadatacenter/cedar-template-editor#813]
We currently validate that a numeric field does not contain invalid characters but we do not validate that the actual values fit within the value space of the associated datatype.

For example, here is a numeric field with a long value:

<img width="797" alt="screen shot 2018-09-05 at 2 21 50 pm" src="https://user-images.githubusercontent.com/1731323/45121946-61531880-b117-11e8-97de-45da6d1427db.png">

It should dynamically be indicated as invalid when it goes outside the long value space.

Ditto for other numeric types.

**Comment by johardi** [comment]
tl;dr;
(a) Implementing a value space for `xsd:int` is possible, (b) implementing a value space for `xsd:float` and `xsd:double` types is also possible but with a critical limitation note for using the "safe" value space, and (c) it is *not* possible to implement the value space for the `xsd:long` type.

I would also recommended to use a string type when materializing number values and transform them to the appropriate number type when it is neccessary (e.g., performing some math operations)

Long version:
Javascript, as well as HTML5, numbers are stored and processed in a [double-precision floating-point](http://www.ecma-international.org/ecma-262/9.0/index.html#sec-ecmascript-language-types-number-type) encoding and therefore they both have a limitation for handling big numbers, especially the `long` number type.

The double-precision floating-point format has 53 bits (2^53) for storing an integer without losing any precision and therefore can be used it to check the value space of `xsd:int`, `xsd:float` and `xsd:double`, such that:
- The `xsd:int` has a value space of [-2^31, 2^31 - 1] or [-2,147,483,648, 2,147,483,647]
- The `xsd:float` has a "safe" value space of [-2^23, 2^23] or [-16,777,216, 16,777,216]
- The `xsd:double` has a "safe" value space of [-2^53, -2^53] or [-9,007,199,254,740,992, 9,007,199,254,740,992]

Note that for the `xsd:float` and `xsd:double` types, when an integer number value goes beyond the "safe" value space, the value representation is no longer accurate, such it will be in the multiplication of 2, 4, 8, etc. depending on the use of the extra bits. 

e.g., 
Typing the number `9,007,199,254,740,997` (which uses 54 bits) in Javascript will be stored as `9,007,199,254,740,996`

An interesting fact is that the number `9,007,199,254,740,997` is still in the space value of `xsd:double` but the accuracy is no longer guaranteed (i.e., not safe)

Also, bear in mind that representing decimal numbers in any floating-point formats will never be accurate, unless they are in a base-2 counting, and the number of digits after the decimal period sign (the precision) is limited depending on the floating-point encoding (e.g., double-precision has 15-17 decimal digits)

References:
* http://www.ecma-international.org/ecma-262/9.0/index.html#sec-ecmascript-language-types-number-type
* https://en.wikipedia.org/wiki/Double-precision_floating-point_format
* https://en.wikipedia.org/wiki/Single-precision_floating-point_format
* https://www.w3.org/TR/xmlschema-2

**Comment by martinjoconnor** [comment]
I also agree with the literal itself as a string. 

In Java there are many options: https://www.baeldung.com/java-check-string-number

Here, for example, is my code in the SWRLAPI for validating some OWL literals:

https://github.com/protegeproject/swrlapi/blob/master/src/main/java/org/swrlapi/literal/OWLLiteralValidator.java

Perhaps we could do something with regular expressions to validate that a literal looks like a number and then using range checks for the specific types (e.g., for xsd:int comparing with the Java constants Integer.MIN_VALUE, Integer.MAX_VALUE). Worth noting that xsd:int and xsd:integers have very different values spaces. Ditto for xsd:double and xsd:decimal.

**Comment by martinjoconnor** [comment]
This is always useful:

![download](https://user-images.githubusercontent.com/1731323/48737256-6d248600-ec02-11e8-9c9f-33cc5517055a.jpg)

**Comment by johardi** [comment]
So @willrett implemented the [numeric form field](https://github.com/metadatacenter/cedar-template-editor/blob/develop/app/scripts/form/runtime-field/numeric.html) as "<input type=number" ...> (line 13)  in the instance editor

Pros:
- It has an inbuilt HTML5 function to check if the input value is within the double-precision FP range.
- The inbuilt function will also check if the input string can be parsed to a valid number representation.

Cons:
- The inbuilt function doesn't discriminate integer-typed numbers because everything is stored in double-precision FP in Javascript.

I can create a new validation function to fully check the `xsd:int` and partially check the `xsd:long`
